src/__tests__/end-to-end.js

```diff
@@ -1,73 +1,150 @@
 import * as React from 'react'
 import {render, waitForElementToBeRemoved, screen, waitFor} from '../'
 
-const fetchAMessage = () =>
-  new Promise(resolve => {
-    // we are using random timeout here to simulate a real-time example
-    // of an async operation calling a callback at a non-deterministic time
-    const randomTimeout = Math.floor(Math.random() * 100)
-    setTimeout(() => {
-      resolve({returnedMessage: 'Hello World'})
-    }, randomTimeout)
-  })
-
-function ComponentWithLoader() {
-  const [state, setState] = React.useState({data: undefined, loading: true})
-  React.useEffect(() => {
-    let cancelled = false
-    fetchAMessage().then(data => {
-      if (!cancelled) {
-        setState({data, loading: false})
-      }
+describe.each([
+  ['real timers', () => jest.useRealTimers()],
+  ['fake legacy timers', () => jest.useFakeTimers('legacy')],
+  ['fake modern timers', () => jest.useFakeTimers('modern')],
+])(
+  'it waits for the data to be loaded in a macrotask using %s',
+  (label, useTimers) => {
+    beforeEach(() => {
+      useTimers()
+    })
+
+    afterEach(() => {
+      jest.useRealTimers()
     })
 
-    return () => {
-      cancelled = true
+    const fetchAMessageInAMacrotask = () =>
+      new Promise(resolve => {
+        // we are using random timeout here to simulate a real-time example
+        // of an async operation calling a callback at a non-deterministic time
+        const randomTimeout = Math.floor(Math.random() * 100)
+        setTimeout(() => {
+          resolve({returnedMessage: 'Hello World'})
+        }, randomTimeout)
+      })
+
+    function ComponentWithMacrotaskLoader() {
+      const [state, setState] = React.useState({data: undefined, loading: true})
+      React.useEffect(() => {
+        let cancelled = false
+        fetchAMessageInAMacrotask().then(data => {
+          if (!cancelled) {
+            setState({data, loading: false})
+          }
+        })
+
+        return () => {
+          cancelled = true
+        }
+      }, [])
+
+      if (state.loading) {
+        return <div>Loading...</div>
+      }
+
+      return (
+        <div data-testid="message">
+          Loaded this message: {state.data.returnedMessage}!
+        </div>
+      )
     }
-  }, [])
 
-  if (state.loading) {
-    return <div>Loading...</div>
-  }
+    test('waitForElementToBeRemoved', async () => {
+      render(<ComponentWithMacrotaskLoader />)
+      const loading = () => screen.getByText('Loading...')
+      await waitForElementToBeRemoved(loading)
+      expect(screen.getByTestId('message')).toHaveTextContent(/Hello World/)
+    })
+
+    test('waitFor', async () => {
+      render(<ComponentWithMacrotaskLoader />)
+      // eslint-disable-next-line testing-library/prefer-find-by -- Sir, this is a test.
```

Comment:
Lol, in fact I think that if `testing-library/prefer-find-by` errored against this, it's a problem because we shouldn't use `findBy` inside a `waitFor`.. Wdyt?

Reply:
These files test Testing Library. They're not meant as documentation for "how to write tests". This is just a temporary fix. I'll probably just remove the plugin since we're not writing tests against an application/library (where these rules do make sense) but against React Testing Library itself.

The "test" I refer to here, is not the "test" you write against an app/library. The test you write against an app/library is considered "production" code from our perspective.

It's like enabling Rule of Hooks for a test that asserts on error messages if Rules of Hooks are vioalted. Or writing tests that verify runtime type-checking in TypeScript.

Reply:
Again, I wasn't arguing that there's an issue with what you're doing, I was just saying that if the plugin errored in this case, it's wrong because inside a 'waitFor' we should use getBy and not findBy.

Reply:
> it's wrong because inside a waitFor we should use getBy and not findBy.

That's not what the warning is suggesting though. It wants me to use `findBy` instead of `waitFor` + `getBy`. But it's not relevant because I specifically want to test that combination. It's kind of like a test for internals so there's no reason to apply production-level lint rules.

It's like linting against using the `arguments` object when my test specifically wants to test interactions with `arguments`.

Reply:
> It's kind of like a test for internals so there's no reason to apply production-level lint rules

I'm 100% on board with you.
I'm resolving this one :)